KeenClient/KeenClient.m

```diff
@@ -291,78 +228,140 @@ + (BOOL)validateKey:(NSString *)key {
     return [KeenClient validateProjectID:key];
 }
 
-- (id)initWithProjectID:(NSString *)projectID andWriteKey:(NSString *)writeKey andReadKey:(NSString *)readKey {
+- (instancetype)initWithNetwork:(KIONetwork*)network
+                       andStore:(KIODBStore*)store
+                    andUploader:(KIOUploader*)uploader {
+    self = [super init];
+
+    if (nil != self) {
+        // log the current version number
+        if ([KeenClient isLoggingEnabled]) {
+            KCLog(@"KeenClient-iOS %@", kKeenSdkVersion);
+        }
+
+        self.network = network;
+        self.store = store;
+        self.uploader = uploader;
+
+        [self refreshCurrentLocation];
+
+        // use global concurrent dispatch queue to run queries in parallel
+        self.queryQueue = dispatch_queue_create(DISPATCH_QUEUE_PRIORITY_DEFAULT, 0);
+    }
+
+    return self;
+}
+
+- (id)initWithProjectID:(NSString *)projectID
+            andWriteKey:(NSString *)writeKey
+             andReadKey:(NSString *)readKey
+             andNetwork:(KIONetwork*)network
+               andStore:(KIODBStore*)store
+            andUploader:(KIOUploader*)uploader {
+    // Validate key parameters
     if (![KeenClient validateProjectID:projectID]) {
+        KCLog(@"Invalid projectID: %@", projectID);
         return nil;
     }
 
-    if (!dbStore) {
-        dbStore = [[KIODBStore alloc] init];
+    if (nil != writeKey) {
+        // only validate a non-nil value
+        KCLog(@"Invalid writeKey: %@", writeKey);
```

Comment:
Why is this log message outside the if(!validateKey...) check? Same question below for readKey.